src/string/aho_corasick.cpp

```diff
@@ -0,0 +1,77 @@
+/**
+ * Author: Richard Hladík
+ * Description: A (non-compressed) trie implementation and the Aho-Corasick
+ *  algorithm. \verb|occurence(i, id)} is called every time there is an
+ *  occurence of the needle with id {\tt id} ending at position $i$.
+ * Time: $O(S + J + V)$, where $S$ is the length of the haystack, $J$ sum of
+ *  the needles' lengths and $V$ is the total number of occurences.
+ * Usage: string s = "cococonut";
+ *  vector<string> n = {"coc", "con"};
+ *  AC(s, n);
+ */
+#include "../base.hpp"
+
+inline int normalize(char c) { return c - 'a'; }
+inline int countBits(int a) { return __builtin_popcount(a); }
+void occurence(int i, int id);
+
+struct Trie {
+	ll bitmask = 0;  // i-th bit is set IFF there is an edge "'a' + i"
+	vector <Trie *> sons = {};  // sorted in alphabetical order
+	int end_of = -1;  // index of the needle ending here or -1
+	Trie *back = NULL, *output = NULL;  // edges
+	char letter = '\0';  // the letter on the edge from parent to here
+
+	Trie (char c) : letter(c) {}
+	~Trie() { for (auto a: sons) delete a; }
+	inline bool hasChild(char c) { return bitmask & (1 << normalize(c)); }
+	inline int childIndex(char c) { return countBits(bitmask % (1 << normalize(c))); }
+	inline void createChild(char c) {
+		sons.insert(sons.begin() + childIndex(c), new Trie(c));  // maintain ordering
```

Comment:
It feels surprising to give up speed for memory efficiency, esp. in a competitive programming context
(It would've been simpler to fill all of the vectors from the start and drop the bit magic)

Reply:
Actually, I think it is *quicker* than prepopulating the vectors, since sons.insert will be called only once per needle insertion, which means O(N + |Σ|) per insert. With resizes you would get O(N|Σ|), unless vector.resize is implemented lazily (which, as far as I can tell from simple for-loop experiments, isn't the case). Of course, set/unordered_set could also work but I don't think the constant is worth it.